### PR TITLE
Display read receipts: list of people who've read the message

### PR DESCRIPTION
**Pull request by Sushmey**
**What does this PR do?**  
As of ZFL 139, one can view all the users who have read a message. This commit allows users to see the read receipts on the message through the message information popup window where it opens another popup displaying all the names of users who've read the message.

New feature #1247 #1268

**Tested?** 
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)

**Commit flow** 
- First commit : Adding the function which fetches the list of users who've read the message from the endpoint
- Second commit : Adding a popup that displays the list
- Third commit : Accounting for different ZFLs and displaying messages accordingly. Also displaying the list of users in a sorted fashion, similar to StreamMembersView.

**Notes & Questions** 

**Interactions** 

**Visual changes** 
The read receipts can be accessed from the message information popup window 
Screenshot 2024-02-08 at 11 45 32 PM

Screenshot 2024-02-08 at 11 46 00 PM

### CODE REVIEW

**Review by mounilKshah — COMMENTED** [review]
Thank you for trying this @Sushmey 👍 
I tried this out on my local machine and it worked just as expected. As for how to display the names of those who have read the message and which key binding to use, we can discuss that further on the channel.

Apart from this, maybe you can modify the commit structure into two commits, with the `Model.fetch_read_receipts()` in the first commit. It would be helpful if you can also add a test function for the same and put that as well in the first commit. And in the second commit, you can add the `ReadReceiptView` class and other functions which handle the UI.

**Review by Kasbr305 — APPROVED** [review]
Fo

File: zulipterminal/model.py [hunk 1]
```diff
@@ -825,6 +825,20 @@ def fetch_message_history(
         display_error_if_present(response, self.controller)
         return list()
 
+    def fetch_read_receipts(self, message_id: int) -> Optional[List[int]]:
+        """
+        Fetches read receipts of a message using its ID.
+        """
```
Comment (neiljp) [inline zulipterminal/model.py:831]: I understand the endpoint is termed read receipts, and that is not a bad name for this function, but internally the result we obtain is a list of user ids. It would be more readable if this method was phrased to indicate that. This avoids the question 'what is a read receipt'?

This and a test for this would be good in a separate first/early commit.

File: zulipterminal/ui_tools/views.py [hunk 2]
```diff
@@ -1929,6 +1942,57 @@ def keypress(self, size: urwid_Size, key: str) -> str:
...
+    ) -> None:
+        self.controller = controller
+        self.message = message
+        self.topic_links = topic_links
+        self.message_links = message_links
+        self.time_mentions = time_mentions
+        max_cols, max_rows = controller.maximum_popup_dimensions()
+
+        # Get rendered message header and footer
+        msg_box = MessageBox(message, controller.model, None)
```
Comment (neiljp) [inline zulipterminal/ui_tools/views.py:1963]: What is the purpose of this?

File: zulipterminal/model.py [hunk 3]
```diff
@@ -834,7 +834,7 @@ def fetch_read_receipts(self, message_id: int) -> Optional[List[int]]:
             f"/messages/{message_id}/read_receipts", method="GET"
         )
         if response["result"] == "success":
-            return response["user_ids"]
+            return ", ".join(list(map(self.user_name_from_id, response["user_ids"])))
```
Comment (neiljp) [inline zulipterminal/model.py:837]: I appreciate that this simplifies things from the point of view of the UI code, but note the recent issues we've had with users with identical names - if we pass only the names out, then we lose the ability to determine which users are which.

Somewhere we'll need to identify that the names are the same and perform an operation to distinguish between them (eg. a suffix). That is a UI-related operation in that it will involve (conditional) styling, so isn't model/API related.

File: tests/model/test_model.py [hunk 4]
```diff
@@ -1430,6 +1430,38 @@ def test_fetch_raw_message_content(
...
+    def test_fetch_read_receipts(
+        self,
+        mocker,
+        model,
+        expected_read_receipts,
+        response,
+        display_error_called,
+        message_id=1,
+    ):
+        self.client.get_read_receipts = mocker.Mock(return_value=response)
```
Comment (neiljp) [inline tests/model/test_model.py:1457]: In the method you've written, you're not using `Client.get_read_receipts`?
Instead you're using `.call_endpoint`?

**Review by neiljp — COMMENTED** [review]
@Sushmey This looks to be along the right lines, thanks for the contribution :+1:

I've left a few comments inline.

For now I'd stick with the user ids being the values returned from the model method, due to not being able to distinguish users by name in some cases.

To aid with reading/reviewing and testing, I'd suggest splitting into smaller (minimal, coherent) commits, at least the model change first - with test.

**Comment by neiljp** [comment]
@Sushmey Another point to bear in mind, is that as per the API docs, this feature is only present from feature level 137 (look for 'ZFL' in the repo). I'm unsure how servers respond to non-existent endpoints (we could check), however a clean way to ensure we don't fall afoul of this would be to only fetch the receipts if ZFL meets this criteria, and likely don't show the option in the UI if the server doesn't support the feature.

Note in the API docs the change at ZFL 143 also, regarding muted users, which we don't support yet, but may wish to at least fetch the data in order to ensure they're not in the data set shown between 135 and 143.

File: zulipterminal/model.py (hunk 1 above)
Comment (Sushmey) [reply zulipterminal/model.py:831]: Yep that's a good point. Changing it to `Fetches user IDs of users who've read a message using its ID.`

Re the commit structure, I was going to add the model function and test in one commit and the UI part in a separate commit but couldn't figure out the test case error so doing this for now. Will change it to the proper commit structure as soon as I figure out the test case

File: zulipterminal/ui_tools/views.py (hunk 2 above)
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1963]: Ah I realised that we don't need a message header or footer for this popup but forgot to remove that code line. Will do that.

File: zulipterminal/ui_tools/views.py [hunk 5]
```diff
@@ -1929,6 +1942,57 @@ def keypress(self, size: urwid_Size, key: str) -> str:
...
+
+        # Get read receipt user ids list
+        response = controller.model.fetch_read_receipts(message["id"])
+        if response is None:
+            return
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1968]: I wrote the ReadReceiptView class based on the `FullRawMsgView` because it seems the most similar. They've handled it exactly like that.

File: tests/model/test_model.py (hunk 4 above)
Comment (Sushmey) [reply tests/model/test_model.py:1457]: I couldn't find a function that did that. Thought there wasn't one so directly called the endpoint. 
Also, I wrote that so that I could add an attribute to the client with `get_read_receipts` as name and then add the parameterised response as the value hit returns hoping it would help me solve the test cases errors

Comment (neiljp) [reply tests/model/test_model.py:1457]: To be clear, it's absolutely fine to use the `.call_endpoint` function if there isn't a named version available (you might try adding one via a PR in that repo) - and using it more generally will also work.

The problem here is that in the test, you're giving the `client` object a `.get_read_receipts` attribute, and then calling the function you've defined. That assigns a value based on `.call_endpoint`, which returns some kind of default each time, which you then test. Can you see the problem?

File: zulipterminal/model.py (hunk 1 above)
Comment (neiljp) [reply zulipterminal/model.py:831]: Just to confirm, I was also meaning ideally the name of the function, not just the docstring.

File: zulipterminal/ui_tools/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py:1968]: Yes, that does seem the case. You might try testing to see what happens when you disconnect from the internet and then press `i`+`r` (or `l` if it's still set that way). I'd think it would be good to show the user something in this case, and similarly apply a bug fix for the Raw message view.

File: zulipterminal/model.py (hunk 3 above)
Comment (neiljp) [reply zulipterminal/model.py:837]: If this is a standalone bugfix, this could be a separate commit.

File: tests/model/test_model.py [hunk 6]
```diff
@@ -1430,6 +1429,42 @@ def test_fetch_raw_message_content(
...
+        model,
+        expected_read_receipts,
+        response,
+        display_error_called,
+        message_id=1,
+    ):
+        # self.client.fetch_read_receipts = mocker.patch(MODEL+".fetch_read_receipts", return_value="response")
+        result = model.fetch_read_receipts(message_id)
+        assert self.display_error_if_present.called == display_error_called
+        assert result == expected_read_receipts
```
Comment (neiljp) [inline tests/model/test_model.py:1459]: Let's refer to the code you're testing:
```python3
 response = self.client.call_endpoint(
            f"/messages/{message_id}/read_receipts", method="GET"
        )
 if response["result"] == "success":
     return ", ".join(list(map(self.user_name_from_id, response["user_ids"])))
```
This means the test tries to call the function `self.client.call_endpoint` in the model. However, the `model` you're using here is the model *fixture*, and if you check the client object in there, you'll see it's a mock (`MagicMock` actually). I'd suggest you try that in one of the model tests by print'ing out it and some other elements.

Since the client mock is not set up to respond a certain way to 'call_endpoint', it just creates it as it needs it. That's why the test behaves the same way in each parametrize case.

If you set it explicitly like in the test just below:
```python3
model.client.update_subscription_settings.return_value = response
```
but appropriately for call_endpoint, that should fix your issue.

File: zulipterminal/model.py (hunk 1 above)
Comment (Sushmey) [reply zulipterminal/model.py]: Would you have any suggestions for the name? Something like `fetch_message_reader_list` or `fetch_message_read_list`?

File: zulipterminal/ui_tools/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1964]: `i`+`r` wouldn't work because pressing `i` closes the info popup immediately. I've changed it to `ctrl`+`r`. So final keypress sequence would be `i` to open the message information popup and then `ctrl`+`r`.


Re the uninitialised object, I'm thinking of showing an error message in the response instead of the list of people who've read. If that looks okay will open a separate PR for `FullRawMsgView`

**Review by neiljp — COMMENTED** [review]
@Sushmey Thanks for pushing this forward :+1:

Other than the code comments, my main point would be to improve your commit text style. Functionally it looks close except for the final styling, which might benefit from combining with the other view I mention? It doesn't show the user ids, if identical names exist, which is maybe dependent on that other PR.

I understand that the lower ZFL or no-connection issues were later additions, but I think it'd be useful to incorporate at least the former into the first commit.

The UI commit is very large, and I would normally prefer to split these to separate the new popup code + test from the code that hooks it together later. This has always been an issue with introducing the keys first (which end up in the hotkeys document and help menu) in order to enable them to be used in the popup, and then in the code that connects things together.

This is also worse now since the message information popup and related popups pass so much data around - it would really help if we handled this a different way (links, time mentions, ...). Similarly, overall it feels like we have a lot of test copy/paste which we should simplify for popups.

File: zulipterminal/model.py (hunk 1 above)
Comment (neiljp) [reply zulipterminal/model.py]: This is fine for now, though you seem to have changed the return type?

File: zulipterminal/ui_tools/views.py [hunk 7]
```diff
@@ -1961,9 +1961,23 @@ def __init__(
         # Get read receipt user ids list
         response = controller.model.fetch_read_receipts(message["id"])
         if response is None:
-            return
+            response = [
+                ("msg_bold", "RESPONSE CANNOT BE FETCHED AT THE MOMENT. "),
+                "\n\nPLEASE TRY AGAIN LATER",
+            ]
+            max_cols = 50
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1968]: Yeah it was the same for me when I tried it. Thought this was a separate issue (perhaps a more generalised solution) so didn't pursue it further and decided to focus on the uninitialised object part by handling the `None` response case. 

I'll look into it trying to see where the code is and solve that first since it might also help in other cases where response is called.

File: tests/model/test_model.py [hunk 8]
```diff
@@ -1487,17 +1487,28 @@ def test_fetch_raw_message_content(
...
                 None,
                 True,
+                137,
+                id="valid_ZFL_with_invalid_response",
+            ),
+            case(
+                {"msg": "", "result": "success", "user_ids": [3, 7, 9]},
+                "unsupported_ZFL",
+                False,
+                12,
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply tests/model/test_model.py:1510]: Alright I'll change the ZFL to some edge case like 136 perhaps? Just under the permissible ZFL. 3 cases seemed perfect - `valid_ZFL_with_valid_response`, `valid_ZFL_with_invalid_response`, `invalid_ZFL_with_response` and I don't think more are necessary

File: zulipterminal/model.py [hunk 9]
```diff
@@ -833,6 +833,20 @@ def fetch_message_history(
         display_error_if_present(response, self.controller)
         return list()
 
+    def fetch_read_receipts(self, message_id: int) -> Optional[str]:
+        """
+        Fetches user IDs of users who've read a message using its ID.
+        """
+        # Get read receipts for a message
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/model.py:840]: Will remove 👍

File: zulipterminal/ui_tools/views.py [hunk 10]
```diff
@@ -1961,9 +1961,23 @@ def __init__(
...
+        elif response == "unsupported_ZFL":
+            response = [
+                (
+                    "msg_bold",
+                    "YOUR CURRENT ZT VERSION DOESN'T SUPPORT VIEWING READ RECEIPTS. ",
+                ),
+                "\n\nPLEASE UPDATE AND TRY AGAIN LATER",
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1975]: I might need some help figuring out how to disable the option dependent on the server version, perhaps file where the specific logic resides? I tried `git grep ZFL` to see where other ZFL based logic was implemented but couldn't find anything similar.

File: zulipterminal/ui_tools/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py:1968]: Let's focus on the rest of the PR first, though you might open an issue for the cases we have here to avoid losing track of them.

File: docs/hotkeys.md [hunk 11]
```diff
@@ -62,6 +62,7 @@
 |View current message in browser (from message information)|<kbd>v</kbd>|
 |Show/hide full rendered message (from message information)|<kbd>f</kbd>|
 |Show/hide full raw message (from message information)|<kbd>r</kbd>|
+|View read receipts of message|<kbd>ctrl</kbd> + <kbd>r</kbd>|
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply docs/hotkeys.md:65]: Changing it to `Show/hide read receipts of message (from message information)`

File: zulipterminal/model.py [hunk 12]
```diff
@@ -838,13 +838,15 @@ def fetch_read_receipts(self, message_id: int) -> Optional[str]:
...
+        if self.server_feature_level is not None and self.server_feature_level >= 137:
+            response = self.client.call_endpoint(
+                f"/messages/{message_id}/read_receipts", method="GET"
+            )
+            if response["result"] == "success":
+                return response["user_ids"]
+            display_error_if_present(response, self.controller)
+        else:
+            return "unsupported_ZFL"
         return None
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/model.py]: Yep, added that as a return type. Didn't know we could annotate multiple return types like this. Thanks!

File: zulipterminal/model.py (hunk 1 above)
Comment (timabbott) [reply zulipterminal/model.py]: I would use a name like `fetch_messaged_read_receipt_user_ids` for this sort of thing -- but I'm particularly a fan of names that make the expected types clear, and not particularly focused on avoiding overly long names :)

**Comment by neiljp** [comment]
@Sushmey With this extra commit this looks better; you said in the stream this was a WIP commit so I'll wait to review further and confirm if you've addressed the points, and handled the mocking OK?

**Comment by Sushmey** [comment]
@zulipbot add "PR needs review"

**Comment by Sushmey** [comment]
@zulipbot remove label "PR awaiting update"

**Review by neiljp — COMMENTED** [review]
@Sushmey Thanks for the update 👍 

It took me a while to recall where we were with this PR, but certainly with the latest server it looks to work well :)

I did notice that we lose the header if we scroll, and we could style it differently - maybe as a section/category header, in the short term, like in other popups? (unless we can put it into the title?)

My inline comments are focused on the flow around the new model method: essentially limiting calls to it, to when it makes sense to do so. Once you've adjusted code/tests to that, other code that looks a little awkward should be able to be simplified. I hope my comments explain how that should work - I think I touched on these ideas before?

In terms of the commit structure, it looks like the current 3rd commit fixes the 2nd - and the 2nd is quite large? I'd suggest focusing on the other changes first, since the UI code may change a little, but then splitting the combination of the 2nd and 3rd into one that adds the `ReadReceiptView` and one that connects things together (see https://github.com/zulip/zulip-terminal/pull/1287#pullrequestreview-1376814871)

File: zulipterminal/model.py [hunk 13]
```diff
@@ -854,6 +854,22 @@ def fetch_message_history(
         display_error_if_present(response, self.controller)
         return list()
 
+    def fetch_read_receipts(self, message_id: int) -> Union[None, str, List[int]]:
+        """
+        Fetches user IDs of users who've read a message using its ID.
+        If fails to fetch due to being unable to retrieve ZFL, returns None
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/model.py:860]: Changing the name to `fetch_messaged_read_receipt_user_ids`, don't know how I missed that. 

Yes the last part is up to date. If it somehow falls through the if else ladder i.e. if ZFL is neither none or >=137 nor then it returns none. I put that there for obscure edge cases. Am I missing something?

File: tests/model/test_model.py [hunk 14]
```diff
@@ -1528,6 +1528,48 @@ def test_fetch_raw_message_content(
...
         assert return_value == expected_raw_content
 
+    @pytest.mark.parametrize(
+        "response, expected_read_receipts, display_error_called, feature_level",
+        [
+            case(
+                {"msg": "", "result": "success", "user_ids": [3, 7, 9]},
+                [3, 7, 9],
+                False,
+                139,
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply tests/model/test_model.py:1538]: Sorry, wrote 139 as the cusp version because read it in the original issue which I based this PR on. Changing the text in commit and PR to 137.

The test case is meant for 139 though, that was intentional.

File: zulipterminal/ui_tools/views.py [hunk 15]
```diff
@@ -1589,6 +1592,7 @@ def __init__(
                 ("Open in web browser", view_in_browser_keys),
                 ("Full rendered message", full_rendered_message_keys),
                 ("Full raw message", full_raw_message_keys),
+                ("View read receipts", read_receipt_keys),
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1595]: Thank you so much for pointing this out! Just added the condition that displays the button only when ZFL>=137.

File: zulipterminal/model.py [hunk 16]
```diff
@@ -854,6 +854,22 @@ def fetch_message_history(
...
+        """
+        if self.server_feature_level is not None and self.server_feature_level >= 137:
+            response = self.client.call_endpoint(
+                f"/messages/{message_id}/read_receipts", method="GET"
+            )
+            if response["result"] == "success":
+                return response["user_ids"]
+            display_error_if_present(response, self.controller)
+        else:
+            return "unsupported_ZFL"
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/model.py:870]: Thank you for pointing this out! I've made the changes and added necessary changes to the tests.

File: zulipterminal/model.py [hunk 17]
```diff
@@ -854,6 +854,22 @@ def fetch_message_history(
...
+        Fetches user IDs of users who've read a message using its ID.
+        If fails to fetch due to being unable to retrieve ZFL, returns None
+        """
+        if self.server_feature_level is not None and self.server_feature_level >= 137:
+            response = self.client.call_endpoint(
+                f"/messages/{message_id}/read_receipts", method="GET"
+            )
+            if response["result"] == "success":
+                return response["user_ids"]
+            display_error_if_present(response, self.controller)
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/model.py:868]: Yeah it's not correct to call this here. I copied it from a similar function. Removing it in this commit 👍

File: tests/model/test_model.py [hunk 18]
```diff
@@ -1528,6 +1528,48 @@ def test_fetch_raw_message_content(
...
+            ),
+            case(
+                {"msg": "Invalid message(s)", "result": "error"},
+                None,
+                True,
+                137,
+                id="valid_ZFL_with_invalid_response",
+            ),
+            case(
+                {"msg": "", "result": "success", "user_ids": [3, 7, 9]},
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply tests/model/test_model.py:1549]: Created a separate test function specifically for unsupported or invalid zfl which checks if runtime error is raised or not

File: zulipterminal/ui_tools/views.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1595]: Also added multiple changes to tests since had to check if the option is displayed in lower zfls and also check the popup height tests

File: zulipterminal/ui_tools/views.py [hunk 19]
```diff
@@ -2001,33 +2001,55 @@ def __init__(
...
+                    (None, "PLEASE TRY AGAIN LATER"),
+                ]
+                max_cols = 50
+            elif response == "unsupported_ZFL":
+                response = [
+                    (
+                        "msg_bold",
+                        "YOUR ZT VERSION DOESN'T SUPPORT VIEWING READ RECEIPTS.\n\n",
+                    ),
+                    (None, "PLEASE UPDATE AND TRY AGAIN LATER"),
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:2017]: Hide the option completely 👍 
No need to display a message now :)

File: zulipterminal/model.py [hunk 20]
```diff
@@ -854,6 +854,23 @@ def fetch_message_history(
...
+        """
+        if self.server_feature_level is not None and self.server_feature_level >= 137:
+            response = self.client.call_endpoint(
+                f"/messages/{message_id}/read_receipts", method="GET"
+            )
+            if response["result"] == "success":
+                return response["user_ids"]
+        else:
+            raise RuntimeError("Unsupported server feature level.")
+        return None
```
Comment (neiljp) [inline zulipterminal/model.py:872]: It is a little confusing to place this line here, since one has to trace back to identify which `if` statement might fail to get here, as the previous block is an `if`/`else` pair. For multiple failing situations, this would be more reasonable, but it's better to pair it with the matching `if` if there is only one.

An alternative is to handle the quicker/simpler error case(s) first; this also means the main flow of code is less indented. Since you have test coverage, exploring this should be safe :)

File: zulipterminal/model.py [hunk 21]
```diff
@@ -854,6 +854,23 @@ def fetch_message_history(
         display_error_if_present(response, self.controller)
         return list()
 
+    def fetch_message_read_receipt_user_ids(
+        self, message_id: int
+    ) -> Union[None, str, List[int]]:
```
Comment (neiljp) [inline zulipterminal/model.py:859]: Needs an update?

File: tests/model/test_model.py [hunk 22]
```diff
@@ -1528,6 +1528,57 @@ def test_fetch_raw_message_content(
...
+        result = model.fetch_message_read_receipt_user_ids(message_id)
+        assert result == expected_read_receipts
+
+    @pytest.mark.parametrize(
+        "feature_level",
+        [
+            case(
+                136,
+                id="invalid_ZFL_with_response",
+            ),
```
Comment (neiljp) [inline tests/model/test_model.py:1568]: If there is only one case, then I would normally suggest it be inline instead, often with parameters such as eg. the feature level able to be specified as a default parameter to the test function - and the id is the 'extra' name of the test.

However, I'd suggest you could also cover back to the minimum feature level, which is typically `None` (and will shortly be `0` after a refactor in another PR).

File: tests/model/test_model.py [hunk 23]
```diff
@@ -1528,6 +1528,57 @@ def test_fetch_raw_message_content(
...
+        ],
+    )
+    def test_fetch_message_read_receipt_user_ids_invalid_zfl(
```
Comment (neiljp) [inline tests/model/test_model.py:1571]: Nit: When naming test functions, we've aimed to add a second `_` to separate the function name from the situation we're handling.

I need to document this :)

File: zulipterminal/model.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/model.py]: I'll respond to the edge case point in my next review, otherwise this is good :)

File: zulipterminal/model.py (hunk 20 above)
Comment (Sushmey) [reply zulipterminal/model.py:872]: Just to confirm, you're talking about the `return None` line, right? How bout we just remove it? I think the runtime error takes care of every invalid case. Plus we aren't handling the case when (if) it returns `None`

File: zulipterminal/model.py (hunk 21 above)
Comment (Sushmey) [reply zulipterminal/model.py:859]: Yes, removing the `str`

File: tests/model/test_model.py (hunk 23 above)
Comment (Sushmey) [reply tests/model/test_model.py:1571]: Changing it to `test_fetch_message_read_receipt_user_ids__invalid_zfl`

Should I also change the valid case function also to `test_fetch_message_read_receipt_user_ids__valid_zfl`?

File: tests/model/test_model.py (hunk 22 above)
Comment (Sushmey) [reply tests/model/test_model.py:1568]: Adding a `None` case in the invalid zfl test function.

Also I understand keeping it inline for just one case saves space but it just looks so much cleaner and more consistent this way with id and all, right?

File: tests/model/test_model.py (hunk 23 above)
Comment (Sushmey) [reply tests/model/test_model.py:1571]: I just noticed, not many test functions with invalid/error cases have double underscores. I take it its a new thing?

File: zulipterminal/model.py (hunk 20 above)
Comment (neiljp) [reply zulipterminal/model.py:872]: Yes, that was what I was referring to (unfortunately GitHub doesn't mark the line I click when starting to review inline :/)

File: tests/model/test_model.py (hunk 22 above)
Comment (neiljp) [reply tests/model/test_model.py:1568]: I paused reviewing this PR since I wanted to get #1462 merged first, which will make these values zero instead of `None`.

File: tests/model/test_model.py (hunk 23 above)
Comment (neiljp) [reply tests/model/test_model.py:1571]: I've not done a survey on how much of the existing code has this. We don't have a testing 'style guide' right now - and many early tests were intended to get any testing setup! We've refactored and adjusted things quite a lot :)

This is more of a final point to note, ie. a very small thing - it's more important we get the tests there and functional - but things like making them more readable will help others.

**Comment by neiljp** [comment]
@Sushmey I didn't intend to submit a review, hence no general comments, and I didn't look at the later commits as much. As above, I realized merging #1462 would be useful to this and other PRs.

**Review by neiljp — COMMENTED** [review]
@Sushmey Here's the followup review for the other commits.

This is definitely clearer to review in smaller parts, though you've also changed quite a bit overall since my last local version.

Generally this looks good, and my comments are mainly about the tests, some of which may be new?

Re the addition of tests to cover multiple values of ZFL in the last commit for different heights, note that I'm fairly sure there isn't 'full coverage' right now, since there are many combinations to look at. We could look at a refactor for that to make the code + comments cleaner eg. with a base 'minimum height' and adjustments. However, an alternative may be to do something like I did for testing the StreamInfoView, ie. test based on examining the content itself. We can follow through with my comments inline for now if you like, and discuss in the stream beyond that.

In terms of testing with different feature levels, you've probably noticed that the MsgInfoView constructed by default can only have one feature level. If there are only a few tests that need that, setting up manually can be reasonable, but the way we resolve that with pytest generally is to have a fixture which returns a function, a 'factory'.

File: zulipterminal/ui_tools/views.py [hunk 24]
```diff
@@ -1970,6 +1970,61 @@ def keypress(self, size: urwid_Size, key: str) -> str:
...
+        self.message = message
+        self.topic_links = topic_links
+        self.message_links = message_links
+        self.time_mentions = time_mentions
+        max_cols, max_rows = controller.maximum_popup_dimensions()
+
+        # Get read receipt user ids list
+        message_read_user_ids = controller.model.fetch_message_read_receipt_user_ids(
+            message["id"]
+        )
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:2009]: Removing the None alltogether in the `fetch_message_read_receipt_user_ids` keeping the default response as an empty list.

File: zulipterminal/ui_tools/views.py [hunk 25]
```diff
@@ -1970,6 +1970,61 @@ def keypress(self, size: urwid_Size, key: str) -> str:
...
+        user_names = [
+            controller.model.user_name_from_id(id) for id in message_read_user_ids
+        ]
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:2012]: Extracting `controller.model` to `model`
Seems kind of inconsistent since some functions extract but some don't.

File: tests/ui_tools/test_popups.py [hunk 26]
```diff
@@ -984,6 +1005,10 @@ def mock_external_classes(
             self.controller, "maximum_popup_dimensions", return_value=(64, 64)
         )
         mocker.patch(LISTWALKER, return_value=[])
+
+        # To show options based on ZFL
+        self.controller.model.server_feature_level = int()
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply tests/ui_tools/test_popups.py:1010]: Can you help me understand why it's odd? Since we are setting the value of the variable where we are testing it, I don't see why the value where it's initialised is relevant. What would you rather set the value as ?

File: tests/ui_tools/test_popups.py [hunk 27]
```diff
@@ -1134,6 +1159,44 @@ def test_keypress_full_raw_message(
...
+            size = widget_size(msg_info_view)
+            msg_info_view.keypress(size, key)
+            self.controller.show_read_receipts.assert_called_once_with(
+                message=message_fixture,
+                topic_links=OrderedDict(),
+                message_links=OrderedDict(),
+                time_mentions=list(),
+            )
+        else:
+            self.controller.show_read_receipts.assert_not_called()
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply tests/ui_tools/test_popups.py:1198]: Thank you for pointing it out! I checked and the keypress function is called when outside the first branch because the function is always pressable even when it's not displayed. So added a boolean variable `show_read_receipts_label` in views.py similar to `show_edit_history_label` which is set as `server_version_level >= 137` 

So now the keypress is not called even when it's outside the first branch when zfl<137 because the bool isn't true.

File: zulipterminal/ui_tools/views.py [hunk 28]
```diff
@@ -1719,6 +1727,14 @@ def keypress(self, size: urwid_Size, key: str) -> str:
                 time_mentions=self.time_mentions,
             )
             return key
+        elif is_command_key("READ_RECEIPTS", key):
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1730]: Nope, adding a bool, `show_read_receipt_label` similar to `show_edit_history_label`

File: zulipterminal/ui_tools/views.py [hunk 29]
```diff
@@ -1591,6 +1594,11 @@ def __init__(
                 ("Full raw message", full_raw_message_keys),
             ],
         )
+
+        # Only show 'View read receipts' option if ZFL>=137
+        if controller.model.server_feature_level >= 137:
+            viewing_actions[1].append(("View read receipts", read_receipt_keys))
```
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1600]: Moving it below, also the keys for read_receipts.

[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1600]: That makes sense, adding this comment 

`ZFL >= 137 supports the option to show read receipts only show 'View read receipts' option if ZFL>=137`

File: zulipterminal/model.py (hunk 20 above)
Comment (neiljp) [reply zulipterminal/model.py:872]: What you have here is:
```
if continue_after_quick_check:
  do_network_call()
  if network_call_success:
    return success_value
else: # quick check failed
  raise quick_check_error()
return None # network call error
```
I first suggested the `None` might be clearer paired with the `if network_call_success` in a matching `else`, ie. moving the last line in this simplified example just after the fourth.

The alternative of checking for quick errors first is more like:
```
if not continue_after_quick_check:
  raise quick_check_error() # quick check failed

do_network_call()
if network_call_success:
  return success_value
return None # network call error (could be in an 'else')
```

File: tests/ui_tools/test_popups.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply tests/ui_tools/test_popups.py:1010]: Right, you *are* setting the value where it is specifically being used in tests, so that it finds it :+1:

If not specifically being used, does zero make sense? That is, do other tests need the variable, and if so, is zero a good value, or 135, or 137, etc? If it is necessary to explain that, a comment could help.

Regarding being explicit, I meant that `int()` is less clear than simply `0`. Also that a small comment could explain the need for a specific value.

File: zulipterminal/ui_tools/views.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py:1600]: Did you mean to only include the first part of that last line?
(the last part is the same as the original comment)

File: zulipterminal/ui_tools/views.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (neiljp) [reply zulipterminal/ui_tools/views.py:2012]: Right, this often comes down to readability right now, ie. it's a lot easier to use `model.` as a shorthand for `controller.model.` if it's going to be used repeatedly. Similarly one could use just `m.`, but there's a balance to be struck :)

File: zulipterminal/ui_tools/views.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (Sushmey) [reply zulipterminal/ui_tools/views.py:1600]: Yes. The first part explains gives context to the condition and the second part summarises the code below it and explains what the code intends to do at a glance. Is that not a good way of commenting?
